### PR TITLE
Implement avl_ndel()

### PR DESCRIPTION
**Pull request by neithernut**
This PR completes PR #145.

### CODE REVIEW

File: src/libreset/avl.c [hunk 1]
```diff
@@ -255,7 +255,9 @@ avl_ndel(
     void* etc,
     struct r_set_cfg* cfg
 ) {
-    return delete_elements_by_predicate(&el->root, pred, etc, cfg);
+    unsigned int retval= delete_elements_by_predicate(&el->root, pred, etc, cfg);
```
Comment (matthiasbeyer) [inline src/libreset/avl.c]: This shouldn't have passed the pre-commit hook. Missing space around `=` here.

File: src/libreset/avl.c [hunk 2]
```diff
@@ -559,3 +561,41 @@ find_node(
...
+    struct r_set_cfg* cfg
+) {
+    avl_dbg("Removing elements with predicate", NULL);
+
+    // check whether the subtree is empty
+    if (!*root) {
+        return 0;
+    }
+
+    unsigned int retval=0;
```
Comment (matthiasbeyer) [inline src/libreset/avl.c]: Same here.

File: src/libreset/avl.c [hunk 3]
```diff
@@ -559,3 +561,41 @@ find_node(
...
 }
 
+static unsigned int
+delete_elements_by_predicate(
+    struct avl_el** root,
+    r_predf pred,
+    void* etc,
+    struct r_set_cfg* cfg
+) {
+    avl_dbg("Removing elements with predicate", NULL);
```
Comment (matthiasbeyer) [inline src/libreset/avl.c]: Do we actually compile wil `DEBUG` set? So does this compile?

**Comment by matthiasbeyer** [comment]
If you're sure about the upper comment, ACK for rebasing. Please ensure that it compiles!

File: src/libreset/avl.c (hunk 3 above)
Comment (neithernut) [inline src/libreset/avl.c]: With `printf` and other va_arg functions, you can spam the stack as much as you want. The function will not care as it only takes information about the arguments from the formatting string.
That's why they cause so many security issues in some programs.
The only thing which could prevent compilation are compiler warnings.
Frankly, if there is no way to print a message without any argument, the macro is buggy.

Comment (matthiasbeyer) [inline src/libreset/avl.c]: Should we 'fix' the macro?

**Comment by neithernut** [comment]
Mh... so much for that... the debug header is broken:
- It does not include `stdio.h`, which causes warnings/errors because of the implicit declaration of fprintf. (this can be solved by including `stdio.h` in line 49 of `debug.h`)
- It tried to print `__LINE__` using `%s` instead of `%d`

Also: a lot of debugging messages cause warnings/errors because of
- `rs_hash` is not an `int`
- missing casts from `struct something*` to `void*`
- too many arguments (caused by this PR)

Glibc (and maybe other libcs as well) ship with a header providing formatting string macros for all the types in `stdint.h` and other headers. Maybe we should use that, if it's availible.

So... what shall we do about this? Fixing the debug-header is ovbiously nothing that would go in this branch. But what should I do about the paramerterless debugging message? Ommit the debugging message completely?

**Comment by matthiasbeyer** [comment]
Remove the debug printing from this PR and open a new issue with all information (and maybe more?) you posted above. Feel free to implement an alternative and create fixup patches (in one big PR).

**Comment by neithernut** [comment]
I'll do so once I'm done with my current stuff.

**Comment by matthiasbeyer** [comment]
:+1:

**Comment by neithernut** [comment]
Done.

**Pull request merged**